File: poeta2.js
//20100505 jorge@jorgechamorro.com
//5000 poemas de 8192 caracteres.
//2.1 segundos.
//2369.7 poemas / segundo.


var puts= require("sys").puts;
var kCuantos= 5e3;
var kLength= 8192;
var ctr= 0;
var now= +new Date();

function rnd (n) { return (n* Math.random()) >>> 0; }

var words= "Con diez cañones por banda viento en popa a toda vela no corta el mar si no vuela un velero bergantín bajel pirata llamado por su bravura el temido en todo el mar conocido del uno al otro confín".toLowerCase().split(" ");

function poeta (length, r, curr, prev, l) {
  r= [];
  l= 0;
  while (l < length) {
    do curr= words[rnd(words.length)]; while (curr === prev);
    r.push(prev= curr);
    l+= prev.length+ 1;
  }
  ctr++;
  return r.join(" ");
}

while (ctr < kCuantos) poeta(kLength);

puts(kCuantos+ " poemas de "+ kLength+ " caracteres.");
puts((now= (+new Date()- now)/1e3).toFixed(1)+ " segundos.");
puts((kCuantos/ now).toFixed(1)+ " poemas / segundo.");

File: poeta2.rb
#!/usr/bin/env ruby -KU
# poeta.rb
# juanfc 2010-04-14
#5000 poemas de 8192 caracteres.
#11.612755 segundos.
#430.561051188973 poemas / segundo.

kCuantos= 5e3.to_i
kLength= 8192
now= Time.now

$words='Con diez cañones por banda viento en popa a toda vela
no corta el mar si no vuela un velero bergantín bajel
pirata llamado por su bravura el temido en todo el mar
conocido del uno al otro confín'.downcase.split

$words_length = $words.length

def poema(length)
        r=[]
        prev=''
        curr=''
        l = 0

        while (l < length) do
           begin
             curr=$words[rand($words_length)]  
           end until (curr != prev)
           
            l = l + curr.length + 1
            r.push(prev=curr)
        end
        
        return r.join(' ')
end

kCuantos.times do
        poema(kLength)
end

puts(kCuantos.to_s + " poemas de "+ kLength.to_s+ " caracteres.");
puts((now=Time.now - now).to_s + " segundos.");
puts((kCuantos/now).to_s+ " poemas / segundo.");


